Introduce ThreadSafeWeakPtr

achristensen07 opened the pull request:
#### 4b51a0df4d5cf653afd99c08091c1887dce58913
<pre>
Introduce ThreadSafeWeakPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=248051">https://bugs.webkit.org/show_bug.cgi?id=248051</a>
rdar://102483501

Reviewed by Chris Dumez.

ThreadSafeWeakPtr is similar to std::weak_ptr, but with a few key differences to make it fit into WebKit better.
The first is that sizeof(ThreadSafeWeakPtr&lt;T&gt;) is sizeof(void*) and, at least in libc++&apos;s implementation,
sizeof(std::weak_ptr&lt;T&gt;) is 2*sizeof(void*).  That is because each std::weak_ptr contains a pointer to the object
and a pointer to the control block containing the strong and weak reference counts.  Instead of that, we give the
object and the control blocks pointers to each other.  By doing that, we get the ability to make a Ref/RefPtr
from only a pointer to the object, so we can continue to use Ref/RefPtr with types that have thread safe weak pointers.

Another difference between ThreadSafeWeakPtr and std::weak_ptr is that ThreadSafeWeakPtr allows you to create a weak
pointer from a raw pointer, which is often used to make a ThreadSafeWeakPtr from *this without an unnecessary
ref/deref cycle.  It does this safely by release asserting that the object is still alive when this happens,
which has half the cost of a ref/deref cycle.

We also support DestructionThread like we do in ThreadSafeRefCounted, and we have double-delete protection like
have have in ThreadSafeRefCountedBase::derefBase by setting ThreadSafeWeakPtrControlBlock.m_object to null when
we delete the object, so that if something makes a Ref to the object during its destructor execution, it won&apos;t
free the memory twice.

This is just the initial introduction of this class.  I still need to make WeakHashSet work with ThreadSafeWeakPtr
and update the 17 more classes that inherit from both ThreadSafeRefCounted and CanMakeWeakPtr.

ThreadSafeWeakPtr is a different kind of smart pointer than WeakPtr because all you can do with it is get a RefPtr,
which requires that the types be reference counted.  We have many types that inherit from CanMakeWeakPtr that are not
reference counted, and we don&apos;t want to change that with this PR.  I could use some crazy SFINAE to make WeakPtr
just return a different type when it&apos;s thread safe, but that is a lot of work for little benefit to get a pointer type
that has the same name but still behaves quite differently internally and in its API, so I decided to make a different
smart pointer type.

* Source/JavaScriptCore/wasm/WasmInstance.h:
(JSC::Wasm::Instance::setMemory):
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::growShared):
(JSC::Wasm::Memory::registerInstance):
* Source/JavaScriptCore/wasm/WasmMemory.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/MainThread.h:
* Source/WTF/wtf/ThreadSafeRefCounted.h:
* Source/WTF/wtf/ThreadSafeWeakPtr.h: Added.
(WTF::ThreadSafeWeakPtrControlBlock::ref const):
(WTF::ThreadSafeWeakPtrControlBlock::deref const):
(WTF::ThreadSafeWeakPtrControlBlock::strongRef const):
(WTF::ThreadSafeWeakPtrControlBlock::strongDeref const):
(WTF::ThreadSafeWeakPtrControlBlock::makeStrongReferenceIfPossible const):
(WTF::ThreadSafeWeakPtrControlBlock::refCount const):
(WTF::ThreadSafeWeakPtrControlBlock::ThreadSafeWeakPtrControlBlock):
(WTF::ThreadSafeWeakPtrControlBlock::WTF_GUARDED_BY_LOCK):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::ref const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::deref const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::refCount const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr):
(WTF::ThreadSafeWeakPtr::ThreadSafeWeakPtr):
(WTF::ThreadSafeWeakPtr::operator=):
(WTF::ThreadSafeWeakPtr::get const):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::idbStorageManager):
* Source/WebKit/NetworkProcess/storage/QuotaManager.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/NavigationSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.mm:
(-[WKSOSecretDelegate initWithSession:]):
(-[WKSOSecretDelegate webViewDidClose:]):
(-[WKSOSecretDelegate webView:didFinishNavigation:]):
(WebKit::PopUpSOAuthorizationSession::initSecretWebView):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.mm:
(WebKit::SOAuthorizationSession::start):
(WebKit::SOAuthorizationSession::continueStartAfterGetAuthorizationHints):
(WebKit::SOAuthorizationSession::complete):
(WebKit::SOAuthorizationSession::presentViewController):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/UIProcess/Notifications/WebNotification.h:
(WebKit::WebNotification::sourceConnection const):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp:
(WebKit::WebNotificationManagerProxy::providerDidShowNotification):
(WebKit::dispatchDidClickNotification):
(WebKit::WebNotificationManagerProxy::providerDidCloseNotifications):
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::ThreadSafeInstanceCounter::ThreadSafeInstanceCounter):
(TestWebKitAPI::ThreadSafeInstanceCounter::~ThreadSafeInstanceCounter):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/256848@main">https://commits.webkit.org/256848@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71fb35719bb5dec1bf9804aa365602285aa9e230

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96998 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106527 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100972 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6494 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35000 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89390 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/103221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102669 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83613 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74784 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87932 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/300 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83377 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/282 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/28422 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/4742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5074 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/86073 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2305 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1516 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/40802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19395 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,202 @@
+        return nullptr;
+    }
+
+    size_t refCount() const
```
Do we need this function for anything? It seems somewhat of limited value for a thread safe refcounted object, since the answer can just change at any time.

The exception I guess is for the `hasOneRef()` case, where if it's true, you can then reason about being the only owner (and it can't change behind your back). Can we just have that function instead?

achristensen07 replied:
refCount and hasOneRef can both change from another thread if weak pointers exist.  I think we should remove both.

mattwoodrow replied:
I think `hasOneRef` should only be true if there were no weak refs, but agree that it's just better not to offer it until we have a compelling need.

mattwoodrow commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,202 @@
+    void strongRef() const
+    {
+        Locker locker { m_lock };
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
```
Is it possible to avoid locking here, if m_strongReferenceCount is atomic?

I guess not, due to the clang threadsafety annotations, and it's likely preferable to have those to prevent bugs, rather than the slight performance win.

achristensen07 replied:
I'm pretty sure we would still need to lock to handle multiple threads creating and destroying weak and strong pointers and converting between the two.

mattwoodrow commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,202 @@
+#include <wtf/MainThread.h>
+#include <wtf/RefPtr.h>
+
+namespace WTF {
```
I think it'd be worth having a little explanation of the class at the top here, particularly to point out that this is the only safe way to have thread-safe refcounting+weakptrs, but does come at the cost of mutex locking for every ref/deref (strong and weak), and some consumers might be better off looking for patterns that don't need this combination.

achristensen07 replied:
I'm planning to add some static_asserts to WeakPtr stuff to make it a compiler error to have a WeakPtr to a ThreadSafeRefCounted type, but I can't add those until the transition completes.  I don't think such a comment would add much.

mattwoodrow replied:
I think the performance overhead of locking is a bit unexpected (since the other refcounting variants don't have that), and consumers might want to reconsider whether they really need WeakPtrs. Explaining that trade-off upfront seems useful to me :)

Big fan of the static_asserts to block the existing (broken) pattern of ThreadSafeRefCounted+CanMakeWeakPtr!

achristensen07 replied:
WTF::Lock is pretty fast, and most of the time it'll just be an atomic operation and a branch.  Even with lots of concurrent threads it'll mostly just be a few spins and you have the lock.  If someone is reconsidering whether they really need ThreadSafeWeakPtr, we don't want them to start storing raw pointers.  I really don't think adding a comment saying "Watch out, this uses locks" will be useful at all.

achristensen07 replied:
After taking Fujii's excellent suggestion, there will never be more than ~10 operations before releasing the lock, so even in the worst case it will only take a few spins to get the lock.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+        : m_object(&object) { }
+
+    mutable Lock m_lock;
+    mutable size_t m_strongReferenceCount WTF_GUARDED_BY_LOCK(m_lock) { 1 };
```
Note that RefCounted and ThreadSafeRefCounted both use `unsigned`. Why are we using size_t instead here?
Would be nice to be consistent.

achristensen07 replied:
I thought about that, but since either way it'll use the same amount of memory when padded (4 pointers size) I decided to be even more safe than ThreadSafeRefCounted in case someone makes 4 billion strong references.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+template<typename T>
+class ThreadSafeWeakPtrControlBlock {
+    WTF_MAKE_NONCOPYABLE(ThreadSafeWeakPtrControlBlock);
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
```
WTF_MAKE_STRUCT_FAST_ALLOCATED is normally for structs. WTF_MAKE_FAST_ALLOCATED would be better here.

cdumez commented on Source/WTF/wtf/MainThread.h:
```diff
@@ -84,6 +84,8 @@ inline void assertIsMainThread() WTF_ASSERTS_ACQUIRED_CAPABILITY(mainThread) { A
 extern NamedAssertion& mainRunLoop;
 inline void assertIsMainRunLoop() WTF_ASSERTS_ACQUIRED_CAPABILITY(mainRunLoop) { ASSERT(isMainRunLoop()); }
 
+enum class DestructionThread { Any, Main, MainRunLoop };
```
`: uint8_t` ?

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+    void ref() const
+    {
+        Locker locker { m_lock };
+        m_weakReferenceCount++;
```
I know there is no difference for this type but I prefer pre-increment in general whenever possible

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+    {
+        Locker locker { m_lock };
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
+        m_strongReferenceCount++;
```
I'd personally prefer pre-increment.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+        Locker locker { m_lock };
+        if (m_object) {
+            // Calling the RefPtr constructor would call strongRef() and deadlock.
+            m_strongReferenceCount++;
```
pre-increment would be nice.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+
+private:
+    template<typename, DestructionThread> friend class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr;
+    ThreadSafeWeakPtrControlBlock(T& object)
```
explicit?

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+
+    mutable Lock m_lock;
+    mutable size_t m_strongReferenceCount WTF_GUARDED_BY_LOCK(m_lock) { 1 };
+    mutable size_t m_weakReferenceCount WTF_GUARDED_BY_LOCK(m_lock) { 0 };
```
ditto.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+    void ref() const { m_controlBlock.strongRef(); }
+    void deref() const { m_controlBlock.template strongDeref<destructionThread>(); }
+protected:
+    ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr() { }
```
`= default;`

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+class ThreadSafeWeakPtr {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    ThreadSafeWeakPtr() { }
```
`= default;`

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+
+template<typename T>
+class ThreadSafeWeakPtr {
+    WTF_MAKE_FAST_ALLOCATED;
```
I don't think we ever expect to heap-allocate a ThreadSafeWeakPtr? Probably not needed.

achristensen07 replied:
WeakPtr has it. Maybe we should make a macro for "this should never be allocated"

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+    ThreadSafeWeakPtr(ThreadSafeWeakPtr&& other)
+        : m_controlBlock(std::exchange(other.m_controlBlock, nullptr)) { }
+
+    void operator=(ThreadSafeWeakPtr&& other) { m_controlBlock = std::exchange(other.m_controlBlock, nullptr); }
```
Should return a `ThreadSafeWeakPtr&` like assignment operators usually do.

achristensen07 replied:
I thought about that, and that would only enable code like this:
a = WTFMove(b = WTFMove(c))
Whatever.  It is indeed normal C++ to return *this;  Will do.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+
+private:
+    template<typename, DestructionThread> friend class ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr;
+    ThreadSafeWeakPtr(ThreadSafeWeakPtrControlBlock<T>& controlBlock)
```
explicit?

cdumez commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.mm:
```diff
@@ -40,20 +40,20 @@
 {
     if ((self = [super init])) {
-        _session = session;
+        _weakSession = ThreadSafeWeakPtr { session };
```
Ideally the explicit `ThreadSafeWeakPtr { }` wouldn't be needed here.

cdumez commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.mm:
```diff
@@ -202,8 +203,9 @@ static bool isSameOrigin(const WebCore::ResourceRequest& request, const WebCore:
     AUTHORIZATIONSESSION_RELEASE_LOG("continueStartAfterGetAuthorizationHints: Checking page for policy choice.");
-    m_page->decidePolicyForSOAuthorizationLoad(hints, [this, weakThis = WeakPtr { *this }] (SOAuthorizationLoadPolicy policy) {
-        if (!weakThis)
+    m_page->decidePolicyForSOAuthorizationLoad(hints, [this, weakThis = ThreadSafeWeakPtr { *this }] (SOAuthorizationLoadPolicy policy) {
```
We probably don't need to capture `this` here. We should just use strongThis below. Safer.

cdumez commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.mm:
```diff
@@ -316,8 +318,9 @@ static bool isSameOrigin(const WebCore::ResourceRequest& request, const WebCore:
 
-    m_page->websiteDataStore().cookieStore().setCookies(WTFMove(cookies), [this, weakThis = WeakPtr { *this }, response = WTFMove(response), data = adoptNS([[NSData alloc] initWithData:data])] () mutable {
-        if (!weakThis)
+    m_page->websiteDataStore().cookieStore().setCookies(WTFMove(cookies), [this, weakThis = ThreadSafeWeakPtr { *this }, response = WTFMove(response), data = adoptNS([[NSData alloc] initWithData:data])] () mutable {
```
Same comment about capturing `this`

achristensen07 replied:
This one needs this because of AUTHORIZATIONSESSION_RELEASE_LOG

cdumez commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.mm:
```diff
@@ -350,11 +353,12 @@ static bool isSameOrigin(const WebCore::ResourceRequest& request, const WebCore:
             return;
         RELEASE_LOG(AppSSO, "presentViewController: Received NSWindowWillCloseNotification. Dismissing the view controller.");
-        weakThis->dismissViewController();
+        strongThis->dismissViewController();
```
Yes, like this :)

fujii commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,195 @@
+    {
+        bool shouldDeleteControlBlock { false };
+        {
+            Locker locker { m_lock };
```
should we call the destructor without locking `m_lock`? I think it blocks other thread ref/deref-ing  the weak ptr while destructing.

achristensen07 replied:
Good catch!  The destructor call introduces arbitrarily long lock holding times.  I'll move it out.

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,238 @@
+    void strongRef() const
+    {
+        Locker locker { m_lock };
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
```
Aren't we worried about the perf cost of this release assertion?

cdumez commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -0,0 +1,238 @@
+
+        {
+            Locker locker { m_lock };
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(m_object);
```
Aren't we concerned about the cost of this release assertion?

cdumez reviewed (approved):
LGTM

Commit:
Introduce ThreadSafeWeakPtr
https://bugs.webkit.org/show_bug.cgi?id=248051
rdar://102483501

Reviewed by Chris Dumez.

ThreadSafeWeakPtr is similar to std::weak_ptr, but with a few key differences to make it fit into WebKit better.
The first is that sizeof(ThreadSafeWeakPtr<T>) is sizeof(void*) and, at least in libc++'s implementation,
sizeof(std::weak_ptr<T>) is 2*sizeof(void*).  That is because each std::weak_ptr contains a pointer to the object
and a pointer to the control block containing the strong and weak reference counts.  Instead of that, we give the
object and the control blocks pointers to each other.  By doing that, we get the ability to make a Ref/RefPtr
from only a pointer to the object, so we can continue to use Ref/RefPtr with types that have thread safe weak pointers.

Another difference between ThreadSafeWeakPtr and std::weak_ptr is that ThreadSafeWeakPtr allows you to create a weak
pointer from a raw pointer, which is often used to make a ThreadSafeWeakPtr from *this without an unnecessary
ref/deref cycle.  It does this safely by release asserting that the object is still alive when this happens,
which has half the cost of a ref/deref cycle.

We also support DestructionThread like we do in ThreadSafeRefCounted, and we have double-delete protection like
have have in ThreadSafeRefCountedBase::derefBase by setting ThreadSafeWeakPtrControlBlock.m_object to null when
we delete the object, so that if something makes a Ref to the object during its destructor execution, it won't
free the memory twice.

This is just the initial introduction of this class.  I still need to make WeakHashSet work with ThreadSafeWeakPtr
and update the 17 more classes that inherit from both ThreadSafeRefCounted and CanMakeWeakPtr.

ThreadSafeWeakPtr is a different kind of smart pointer than WeakPtr because all you can do with it is get a RefPtr,
which requires that the types be reference counted.  We have many types that inherit from CanMakeWeakPtr that are not
reference counted, and we don't want to change that with this PR.  I could use some crazy SFINAE to make WeakPtr
just return a different type when it's thread safe, but that is a lot of work for little benefit to get a pointer type
that has the same name but still behaves quite differently internally and in its API, so I decided to make a different
smart pointer type.

* Source/JavaScriptCore/wasm/WasmInstance.h:
(JSC::Wasm::Instance::setMemory):
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
(JSC::Wasm::Memory::growShared):
(JSC::Wasm::Memory::registerInstance):
* Source/JavaScriptCore/wasm/WasmMemory.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/MainThread.h:
* Source/WTF/wtf/ThreadSafeRefCounted.h:
* Source/WTF/wtf/ThreadSafeWeakPtr.h: Added.
(WTF::ThreadSafeWeakPtrControlBlock::ref const):
(WTF::ThreadSafeWeakPtrControlBlock::deref const):
(WTF::ThreadSafeWeakPtrControlBlock::strongRef const):
(WTF::ThreadSafeWeakPtrControlBlock::strongDeref const):
(WTF::ThreadSafeWeakPtrControlBlock::makeStrongReferenceIfPossible const):
(WTF::ThreadSafeWeakPtrControlBlock::refCount const):
(WTF::ThreadSafeWeakPtrControlBlock::ThreadSafeWeakPtrControlBlock):
(WTF::ThreadSafeWeakPtrControlBlock::WTF_GUARDED_BY_LOCK):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::ref const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::deref const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::refCount const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr):
(WTF::ThreadSafeWeakPtr::ThreadSafeWeakPtr):
(WTF::ThreadSafeWeakPtr::operator=):
(WTF::ThreadSafeWeakPtr::get const):
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::idbStorageManager):
* Source/WebKit/NetworkProcess/storage/QuotaManager.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/NavigationSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.mm:
(-[WKSOSecretDelegate initWithSession:]):
(-[WKSOSecretDelegate webViewDidClose:]):
(-[WKSOSecretDelegate webView:didFinishNavigation:]):
(WebKit::PopUpSOAuthorizationSession::initSecretWebView):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SOAuthorizationSession.mm:
(WebKit::SOAuthorizationSession::start):
(WebKit::SOAuthorizationSession::continueStartAfterGetAuthorizationHints):
(WebKit::SOAuthorizationSession::complete):
(WebKit::SOAuthorizationSession::presentViewController):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/UIProcess/Notifications/WebNotification.h:
(WebKit::WebNotification::sourceConnection const):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp:
(WebKit::WebNotificationManagerProxy::providerDidShowNotification):
(WebKit::dispatchDidClickNotification):
(WebKit::WebNotificationManagerProxy::providerDidCloseNotifications):
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::ThreadSafeInstanceCounter::ThreadSafeInstanceCounter):
(TestWebKitAPI::ThreadSafeInstanceCounter::~ThreadSafeInstanceCounter):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/256848@main
